stat/card/hll32.go

```diff
@@ -142,3 +146,82 @@ func (h *HyperLogLog32) Reset() {
 		h.register[i] = 0
 	}
 }
+
+// MarshalBinary marshals the sketch in the receiver. It encodes the
+// name of the hash function, the precision of the sketch and the
+// sketch data. The receiver must have a non-nil hash function.
+func (h *HyperLogLog32) MarshalBinary() ([]byte, error) {
+	if h.hash == nil {
+		return nil, errors.New("card: hash function not set")
+	}
+	t := reflect.TypeOf(h.hash)
+	var prefix string
+	if t.Kind() == reflect.Ptr {
+		t = t.Elem()
+		prefix = "*"
+	}
+	var buf bytes.Buffer
+	enc := gob.NewEncoder(&buf)
```

Comment:
Given that this is basically gob, why not call this GobEncoder? I would have thought that a BinaryEncoder would maybe store the states directly to make a more easily readable format.

Reply:
Both are picked up by gob, but other binary encoding paths don't pick up `gob.Gob{Encoder,Decoder}`. If we only provide one serialisation hook, then it makes sense to have it work with all binary encoding paths, `encoding.Binary{Marshaler,Unmarshaler}` gets us this.

Reply:
Okay. I'm not sure that this is right, but I'll admit that I can't think through the long-term ramifications of the decision.

Reply:
The only issue that I can see is that it gets in the way of foreign language interoperability. I don't think this is a huge issue, particularly since the implementation is diverging from what people normally do anyway.